Add Enterprise SONIC LLDP Global module

divya-balasubramania opened the pull request:
##### SUMMARY
LLDP_GLOBAL Module implementation

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
sonic_lldp_global

##### ADDITIONAL INFORMATION

1. Regression test HTML report:  
[sonic_lldp_global_test_results.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/10848594/sonic_lldp_global_test_results.pdf)


2. Regression test detailed terminal log:
[lldp_global_terminal_test.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/10848597/lldp_global_terminal_test.txt)


3. Show running-config and respective facts gathered:
[LLDP_Global_facts_UT.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/10848601/LLDP_Global_facts_UT.txt)

ArunSaravananBalachandran commented on plugins/modules/sonic_lldp_global.py:
```diff
@@ -0,0 +1,301 @@
+    dellemc.enterprise_sonic.sonic_lldp_global:
+      config:
+        tlv_select:
+          system_capabilities: True
```
Please change the value to `true`

ArunSaravananBalachandran commented on plugins/modules/sonic_lldp_global.py:
```diff
@@ -0,0 +1,301 @@
+              mode: receive
+              system_description: Sonic_System
+              tlv_select:
+                  management_address: True
```
Please change the value to `true`

ArunSaravananBalachandran commented on tests/regression/roles/sonic_lldp_global/defaults/main.yml:
```diff
@@ -0,0 +1,42 @@
+            system_capabilities: false
+        
+
+  - name: test_case_02
+    description: Update Global LLDP configuration
+    state: merged
+    input:
+        multiplier: 10 
+        system_name: CR1_sonic
+        hello_time: 200 
+        tlv_select: 
+           system_capabilities: true
```
Please update the indentation to be uniform across blocks for all test cases.

ArunSaravananBalachandran commented on tests/regression/roles/sonic_lldp_global/defaults/main.yml:
```diff
@@ -0,0 +1,42 @@
+        hello_time: 18
+        mode: receive
+        system_description: Sonic_CR1_device
+        tlv_select: 
+            system_capabilities: false
+        
```
Please remove trailing whitespaces in this file, cleanup_tests.yaml, preparation_tests.yaml and tasks_template_del.yaml

ArunSaravananBalachandran commented on plugins/module_utils/network/sonic/config/lldp_global/lldp_global.py:
```diff
@@ -0,0 +1,308 @@
+
+        if not command:
+            return requests
+        if "enable" in command:
+            payload = {'openconfig-lldp:enabled': command['enable']}
```
Replace double quotes with single quote for keys in config and facts.

divya-balasubramania commented:
Addressed review comments and re-executed the test cases.

Regression test HTML report:

[lldp_global_regression_results.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/10956240/regression-2023-03-13-15-19-50.pdf)

Regression test detailed terminal log:

[lldp_global_detailed_log_latest.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/10956248/lldp_global_detailed_log_latest.txt)

ArunSaravananBalachandran commented on plugins/module_utils/network/sonic/facts/lldp_global/lldp_global.py:
```diff
@@ -0,0 +1,113 @@
+            response = edit_config(self._module, to_request(self._module, request))
+        except ConnectionError as exc:
+            self._module.fail_json(msg=str(exc), code=exc.code)
+        if 'openconfig-lldp:config' in response[0][1]:
+            raw_lldp_global_data = response[0][1]['openconfig-lldp:config']
+            if 'enabled' in raw_lldp_global_data:
+                lldp_global_data['enable'] = raw_lldp_global_data['enabled']
+            if 'hello-timer' in raw_lldp_global_data:
+                lldp_global_data['hello_time'] = raw_lldp_global_data['hello-timer']
+            if 'openconfig-lldp-ext:mode' in raw_lldp_global_data:
+                lldp_global_data['mode'] = raw_lldp_global_data['openconfig-lldp-ext:mode'].lower()
+            if 'system-description' in raw_lldp_global_data:
+                lldp_global_data['system_description'] = raw_lldp_global_data['system-description']
+            if 'system-name' in raw_lldp_global_data:
+                lldp_global_data['system_name'] = raw_lldp_global_data['system-name']
+            if 'openconfig-lldp-ext:multiplier' in raw_lldp_global_data:
+                lldp_global_data['multiplier'] = raw_lldp_global_data['openconfig-lldp-ext:multiplier']
+            lldp_global_data['tlv_select'] = {}
+            lldp_global_data['tlv_select']['management_address'] = True
+            lldp_global_data['tlv_select']['system_capabilities'] = True
+            if 'suppress-tlv-advertisement' in raw_lldp_global_data:
+                for tlv_select in raw_lldp_global_data['suppress-tlv-advertisement']:
+                    tlv_select = tlv_select.replace('openconfig-lldp-types:', '').lower()
+                    if tlv_select in ('management_address', 'system_capabilities'):
+                        lldp_global_data['tlv_select'][tlv_select] = False
+        return lldp_global_data
```
Default settings of 'enabled' and 'tlv_select' options needs to be handled.

ArunSaravananBalachandran commented on plugins/modules/sonic_lldp_global.py:
```diff
@@ -0,0 +1,301 @@
+description:
+  - This module provides configuration management of global LLDP parameters
+    for use on LLDP enabled Layer 2 interfaces of devices running SONiC.
+    It is intended for use in conjunction with LLDP Layer 2 interface
+    configuration applied on participating interfaces.
```
Please mention it as a list element to display it in a separate line.

ArunSaravananBalachandran commented on plugins/modules/sonic_lldp_global.py:
Please update the indentation to be uniform across blocks for all examples.

ArunSaravananBalachandran commented on tests/regression/roles/sonic_lldp_global/defaults/main.yml:
```diff
@@ -0,0 +1,41 @@
+      tlv_select:
+        system_capabilities: true
+
+  - name: test_case_03
+    description: Delete specific global lldp configurations
+    state: deleted
+    input:
+      multiplier: 10
+      system_name: CR1_sonic
```
Please include deletion of 'tlv_select' option as mentioned in the module example.

ArunSaravananBalachandran commented on plugins/module_utils/network/sonic/config/lldp_global/lldp_global.py:
```diff
@@ -0,0 +1,308 @@
+            commands = have
+            requests.extend(self.get_delete_lldp_global_completely_requests(commands))
+        else:
+            if "tlv_select" in want and want['tlv_select']:
+                if "management_address" in want['tlv_select']:
+                    if want['tlv_select']['management_address'] is not None and have['tlv_select']['management_address'] is not None:
+                        want['tlv_select']['management_address'] = False
+                if "system_capabilities" in want['tlv_select']:
+                    if want['tlv_select']['system_capabilities'] is not None and have['tlv_select']['management_address'] is not None:
+                        want['tlv_select']['system_capabilities'] = False
```
As stated in the comment (https://github.com/ansible-network/resource_module_models/pull/196#discussion_r1101081735), the deletion needs to be carried out only for matching configurations. 

Usage of diff to get the matching configuration in other options will also work for tlv_select options.
Changing the option value will result in incorrect behaviour.

ArunSaravananBalachandran commented on plugins/module_utils/network/sonic/config/lldp_global/lldp_global.py:
```diff
@@ -0,0 +1,288 @@
+            url = self.lldp_global_config_path['multiplier']
+            requests.append({'path': url, 'method': PATCH, 'data': payload})
+
+        if 'system_name' in command:
+            payload = {'openconfig-lldp:system-name': command['system_name']}
+            url = self.lldp_global_config_path['system_name']
+            requests.append({'path': url, 'method': PATCH, 'data': payload})
+
+        if 'system_description' in command:
```
Please include `None` value check.

ArunSaravananBalachandran commented on plugins/module_utils/network/sonic/config/lldp_global/lldp_global.py:
```diff
@@ -0,0 +1,288 @@
+        if 'system_description' in command:
+            url = self.lldp_global_config_path['system_description']
+            requests.append({'path': url, 'method': DELETE})
+        # The tlv_select configs are enabled by default.Hence false leads deletion of configs.
+        if 'tlv_select' in command:
+            if 'management_address' in command['tlv_select']:
+                if command['tlv_select']['management_address'] is False:
+                    url = self.lldp_suppress_tlv.format(lldp_suppress_tlv="MANAGEMENT_ADDRESS")
+                    requests.append({'path': url, 'method': DELETE})
+            if 'system_capabilities' in command['tlv_select']:
+                if command['tlv_select']['system_capabilities'] is False:
+                    url = self.lldp_suppress_tlv.format(lldp_suppress_tlv="SYSTEM_CAPABILITIES")
+                    requests.append({'path': url, 'method': DELETE})
```
Both `True` and `False` values need to be handled.

divya-balasubramania commented:
Addressed review comments and re-executed the test cases.

Regression test HTML report:
[SONIC_LLDP_Global_Latest.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/11099391/SONIC_LLDP_Global_Latest.pdf)
